src/main/java/fi/fmi/avi/archiver/config/ChannelConfig.java

```diff
@@ -1,68 +1,129 @@
     }
 
     @Bean
     public MessageChannel failChannel() {
-        return new PublishSubscribeChannel(Executors.newSingleThreadExecutor(newThreadFactory("Fail-")));
+        return new PublishSubscribeChannel(failExecutor());
+    }
+
+    @Bean
+    public ExecutorService parserExecutor() {
+        return newBlockingSingleThreadExecutor("Parser-");
     }
 
     @Bean
     public MessageChannel parserChannel() {
-        return new PublishSubscribeChannel(Executors.newSingleThreadExecutor(newThreadFactory("Parser-")));
+        return new PublishSubscribeChannel(parserExecutor());
+    }
+
+    @Bean
+    public ExecutorService populatorExecutor() {
+        return newBlockingSingleThreadExecutor("Populator-");
     }
 
     @Bean
     public MessageChannel populatorChannel() {
-        return new PublishSubscribeChannel(Executors.newSingleThreadExecutor(newThreadFactory("Populator-")));
+        return new PublishSubscribeChannel(populatorExecutor());
+    }
+
+    @Bean
+    public ExecutorService databaseExecutor() {
+        return newBlockingSingleThreadExecutor("Database-");
     }
 
     @Bean
     public MessageChannel databaseChannel() {
-        return new PublishSubscribeChannel(Executors.newSingleThreadExecutor(newThreadFactory("Database-")));
+        return new PublishSubscribeChannel(databaseExecutor());
+    }
+
+    @Bean
+    public ExecutorService finishExecutor() {
+        return newBlockingSingleThreadExecutor("Finish-");
+    }
+
+    @Bean
+    public MessageChannel finishChannel() {
+        return new PublishSubscribeChannel(finishExecutor());
+    }
+
+    @Bean
+    public ExecutorService errorMessageExecutor() {
+        return newBlockingSingleThreadExecutor("Error-");
     }
 
     @Bean
     public MessageChannel errorMessageChannel() {
-        return new PublishSubscribeChannel(Executors.newCachedThreadPool(newThreadFactory("Error-")));
+        return new PublishSubscribeChannel(errorMessageExecutor());
     }
 
     @Bean
-    public MessageChannel errorLoggingChannel() {
-        return new PublishSubscribeChannel(Executors.newCachedThreadPool(newThreadFactory("Error-Log-")));
+    public ExecutorService errorLoggingExecutor() {
+        return Executors.newCachedThreadPool(newThreadFactory("Error-Log-"));
     }
 
     @Bean
-    public MessageChannel finishChannel() {
-        return new PublishSubscribeChannel(Executors.newSingleThreadExecutor(newThreadFactory("Finish-")));
+    public MessageChannel errorLoggingChannel() {
+        return new PublishSubscribeChannel(errorLoggingExecutor());
+    }
+
+    private ExecutorService newBlockingSingleThreadExecutor(final String threadNamePrefix) {
+        final BlockingQueue<Runnable> queue = new ArrayBlockingQueue<>(executorQueueSize);
+        return new ThreadPoolExecutor(1, 1, 0, TimeUnit.SECONDS, queue,
+                newThreadFactory(threadNamePrefix), new CallerBlocksPolicy(Long.MAX_VALUE));
```

Comment:
A shortcoming in the original plan... Now if the whole chain gets stuck for some unknown reason (other than database not responding, for which we are prepared), we get no clear signal of it.

To detect these situations, I'd suggest something like following (you might choose to create a new issue of it):

1. Register each phase executor in a health indicator component (one indicator for all executors).
2. Declare a configurable timeout with a decent default value for blocking.
3. When the timeout is reached, flag executor as unhealthy.
4. When blocking finishes, flag executor as healthy.
5. When all executors are flagged unhealthy, flag application state as unhealthy.

Reply:
👍 Created issue #70.